### PR TITLE
[PM-1977] Display passkeys properly on the desktop

### PR DESCRIPTION
**Pull request by gbubemismith**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Passkeys will be listed in the vault alongside Login items in the browser clients. Anywhere users filter to login item type, passkeys will also be displayed using the same sort order (eg, if a list of login items is sorted by name, the login item “A” will display, then the passkey item “B”, then the login item “C”

When the launch icon is selected on the Application field, launch the top domain of the rpId.

Passkeys item cannot be cloned.

Non-discoverable passkeys = 2FA

Display a passkey field on the Login item above the TOTP field.

Login items with the passkey field can be cloned, but the user is warned that the passkey field will not be included in the cloned item.
## Code changes

- **apps/desktop/src/vault/app/vault/add-edit.component.html:** Added view for `Fido2Key` when editing.
- **apps/desktop/src/vault/app/vault/vault.component.ts:** Added launch option to `Fido2Key` vault item
- **apps/desktop/src/vault/app/vault/view.component.html:** Added vault view for Fido2Key to align with the Figma designs.
-  **libs/angular/src/vault/vault-filter/models/vault-filter.model.ts:** Modified the filter so `Login` and `Fido2Key` would be treated as a single type when building the filter.

## Screenshots
Screenshot 2023-07-07 at 5 26 42 PM
Screenshot 2023-07-07 at 5 26 59 PM
Screenshot 2023-07-07 at 5 27 12 PM
Screenshot 2023-07-07 at 5 27 23 PM

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: apps/desktop/src/vault/app/vault/add-edit.component.html [hunk 1]
```diff
@@ -448,6 +448,24 @@ <h2 class="box-header">
...
+          <div *ngIf="cipher.type === cipherType.Fido2Key">
+            <div class="box-content-row" appBoxRow>
+              <label for="fido2KeyUsername">{{ "username" | i18n }}</label>
+              <input
+                id="fido2KeyUsername"
+                type="text"
+                name="Fido2Key.Username"
+                [(ngModel)]="cipher.fido2Key.userName"
+                appInputVerbatim
+                [readonly]="!cipher.edit && editMode"
```
Comment (coroiu) [inline apps/desktop/src/vault/app/vault/add-edit.component.html:461]: question: aren't passkey usernames readonly? Or am I misremembering things? :)

File: apps/desktop/src/vault/app/vault/add-edit.component.html [hunk 2]
```diff
@@ -448,6 +448,24 @@ <h2 class="box-header">
...
+              />
+            </div>
+            <div class="box-content-row text-muted" appBoxRow>
+              <span class="row-label">{{ "typePasskey" | i18n }}</span>
+              {{ "dateCreated" | i18n }} {{ cipher.creationDate | date : "MM/dd/y" }}
```
Comment (coroiu) [inline apps/desktop/src/vault/app/vault/add-edit.component.html:466]: issue: date format should depend on locale
```suggestion
              {{ "dateCreated" | i18n }} {{ cipher.creationDate | date : "medium" }}
```

File: apps/desktop/src/vault/app/vault/view.component.html [hunk 3]
```diff
@@ -389,6 +396,19 @@ <h2 class="box-header">
...
+        <div *ngIf="cipher.isFido2Key">
+          <div class="box-content-row">
+            <div class="row-main">
+              <span class="row-label">{{ "username" | i18n }}</span>
+              {{ cipher.fido2Key.userName }}
+            </div>
+          </div>
+          <div class="box-content-row">
+            <span class="row-label">{{ "typePasskey" | i18n }}</span>
+            {{ "dateCreated" | i18n }} {{ cipher.creationDate | date : "MM/dd/y" }}
```
Comment (coroiu) [inline apps/desktop/src/vault/app/vault/view.component.html:409]: issue: same here, can we use another date format?

File: apps/desktop/src/vault/app/vault/view.component.html [hunk 4]
```diff
@@ -465,6 +485,45 @@ <h2 class="box-header">
...
+            </span>
+          </div>
+          <div class="action-buttons">
+            <button
+              type="button"
+              class="row-btn"
+              appStopClick
+              (click)="launch(cipher.fido2Key)"
+              attr.aria-label="{{ 'launch' | i18n }} {{ cipher.fido2Key.launchUri }}"
+              [appA11yTitle]="'launch' | i18n"
```
Comment (coroiu) [inline apps/desktop/src/vault/app/vault/view.component.html:509]: issue: this doesn't seem right, `appA11yTitle` directive already sets `aria-label`. Don't use both

File: apps/desktop/src/vault/app/vault/view.component.html [hunk 5]
```diff
@@ -465,6 +485,45 @@ <h2 class="box-header">
...
+            >
+              <i class="bwi bwi-lg bwi-share-square" aria-hidden="true"></i>
+            </button>
+            <button
+              type="button"
+              class="row-btn"
+              appStopClick
+              attr.aria-label="{{ 'copyUri' | i18n }} {{ cipher.fido2Key.launchUri }}"
+              [appA11yTitle]="'copyUri' | i18n"
```
Comment (coroiu) [inline apps/desktop/src/vault/app/vault/view.component.html:518]: issue: same a11y issue

File: libs/angular/src/vault/vault-filter/models/vault-filter.model.ts [hunk 6]
```diff
@@ -45,7 +45,13 @@ export class VaultFilter {
...
       }
       if (this.cipherType != null && cipherPassesFilter) {
-        cipherPassesFilter = cipher.type === this.cipherType;
+        //Fido2Key's should also be included in the Login type
+        if (this.cipherType === CipherType.Login) {
+          cipherPassesFilter =
+            cipher.type === this.cipherType || cipher.type === CipherType.Fido2Key;
+        } else {
+          cipherPassesFilter = cipher.type === this.cipherType;
+        }
```
Comment (coroiu) [inline libs/angular/src/vault/vault-filter/models/vault-filter.model.ts:54]: question: do we have any tests for this?

File: libs/common/src/vault/models/view/cipher.view.ts [hunk 7]
```diff
@@ -131,6 +131,10 @@ export class CipherView implements View, InitializerMetadata {
     return this.item.linkedFieldOptions;
   }
 
+  get isFido2Key() {
+    return Object.keys(this.fido2Key).length > 0;
+  }
+
```
Comment (coroiu) [inline libs/common/src/vault/models/view/cipher.view.ts:137]: issue: this doesn't seem right, why are we checking keys on a class? From personal experience `Object.keys` is also very slow. Can't we just use `this.type === CipherType.Fido2Key` ?

**Review by coroiu — CHANGES_REQUESTED** [review]

File: apps/desktop/src/vault/app/vault/add-edit.component.html (hunk 2 above)
Comment (gbubemismith) [reply apps/desktop/src/vault/app/vault/add-edit.component.html:466]: The date format is based on the Figma designs as well

Comment (coroiu) [reply apps/desktop/src/vault/app/vault/add-edit.component.html:466]: Figma designs usually use `en-US` locale, I've added a comment in Figma to ask about locale-dependent dates

**Review by coroiu — COMMENTED** [review]
👍 just waiting on answer from Danielle here: https://www.figma.com/file/K7u674lSE0UJrAFK9ikzT9?node-id=233:28788&mode=design#494289573

File: apps/desktop/src/vault/app/vault/add-edit.component.html (hunk 2 above)
Comment (coroiu) [reply apps/desktop/src/vault/app/vault/add-edit.component.html:466]: Got a response, you can use the `short` format

Comment (gbubemismith) [reply apps/desktop/src/vault/app/vault/add-edit.component.html:466]: Yeah just saw that. I would be making the changes to the other clients as well

**Review by coroiu — APPROVED** [review]
👍

**Pull request merged**